File: android/app/src/main/kotlin/com/poracode/app/protocol/RuntimeEventReducer.kt
package com.poracode.app.protocol

import com.poracode.app.model.PersistedRuntimeItem
import com.poracode.app.model.asObjectOrNull
import com.poracode.app.model.array
import com.poracode.app.model.obj
import com.poracode.app.model.string
import java.util.UUID
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Unwraps remote supervisor envelopes and reduces runtime events for a thread transcript.
 *
 * TS-equivalent foundation (`runtimeEvent.ts` + iOS RuntimeEventReducer):
 * - unwrap `thread-runtime-event` / `thread-runtime-events` / `thread-runtime-events-multi`
 * - **strict sealed union of 14 variants**; unknown/malformed skipped
 * - `item.updated` requires payload **key presence** (JsonNull distinct from absent)
 * - never fabricate stubs for missing items on updated/completed/delta
 * - empty completed reasoning dropped
 * - interrupted/cancelled turn prunes trailing reasoning (plan/error/child transparent)
 * - synthetic completed error item with injected UUID
 * - request.opened FIFO replace+append; request.resolved removal (via domain reducer)
 * - turn open/closed, context.updated merge, usage.spent intentional no-op
 * - hide `pending_request` transcript rows; recover open requests from them on hydrate
 */
object RuntimeEventReducer {
    data class RuntimeEvent(
        val type: String,
        val threadId: String? = null,
        val itemId: String? = null,
        val itemType: String? = null,
        val state: String? = null,
        val stream: String? = null,
        val delta: String? = null,
        val payload: JsonElement? = null,
        /**
         * True when the wire object contained a `payload` key (including JsonNull).
         * Distinct from payload == null meaning "key absent".
         */
        val payloadSpecified: Boolean = false,
        val parentItemId: String? = null,
        val requestId: String? = null,
        val requestType: String? = null,
        val message: String? = null,
        val raw: JsonObject = JsonObject(emptyMap()),
        /** Populated when parse produced a strict sealed variant. */
        val canonical: RuntimeEventSchema.CanonicalRuntimeEvent? = null,
    )

    data class Batch(
        val threadId: String,
        val events: List<RuntimeEvent>,
    )

    /** Result of applying events: items + domain state mutation. */
    data class ApplyResult(
        val items: List<PersistedRuntimeItem>,
        val domain: ThreadRuntimeDomainState,
    )

    // MARK: - Envelope unwrapping

    fun collectRuntimeEvents(from: JsonElement): List<Batch> {
        val objectMap = from.asObjectOrNull() ?: return emptyList()
        val type = objectMap.string("type") ?: return emptyList()

        return when (type) {
            "thread-runtime-event" -> {
                val threadId = objectMap.string("threadId") ?: return emptyList()
                val eventObject = objectMap.obj("event") ?: return emptyList()
                val event = parseRuntimeEvent(eventObject) ?: return emptyList()
                listOf(Batch(threadId = threadId, events = listOf(event)))
            }

            "thread-runtime-events" -> {
                val threadId = objectMap.string("threadId") ?: return emptyList()
                val rawEvents = objectMap.array("events") ?: return emptyList()
                val events = rawEvents.mapNotNull { value ->
                    value.asObjectOrNull()?.let { parseRuntimeEvent(it) }
                }
                if (events.isEmpty()) emptyList() else listOf(Batch(threadId, events))
            }

            "thread-runtime-events-multi" -> {
                val batches = objectMap.array("batches") ?: return emptyList()
                batches.mapNotNull { batchValue ->
                    val batchObject = batchValue.asObjectOrNull() ?: return@mapNotNull null
                    val threadId = batchObject.string("threadId") ?: return@mapNotNull null
                    val rawEvents = batchObject.array("events") ?: return@mapNotNull null
                    val events = rawEvents.mapNotNull { value ->
                        value.asObjectOrNull()?.let { parseRuntimeEvent(it) }
                    }
                    if (events.isEmpty()) null else Batch(threadId, events)
                }
            }

            else -> emptyList()
        }
    }

    /**
     * Parse a runtime event object against the strict 14-variant sealed union.
     * Unknown / malformed / missing required fields → null (skipped).
     * `item.updated` **requires** payload key presence on the wire (JsonNull distinct).
     */
    fun parseRuntimeEvent(objectMap: JsonObject): RuntimeEvent? {
        val canonical = RuntimeEventSchema.parseCanonical(objectMap) ?: return null
        return RuntimeEventSchema.toRuntimeEvent(canonical).copy(canonical = canonical)
    }

    fun parseCanonical(objectMap: JsonObject): RuntimeEventSchema.CanonicalRuntimeEvent? =
        RuntimeEventSchema.parseCanonical(objectMap)

    fun visibleTranscriptItems(items: List<PersistedRuntimeItem>): List<PersistedRuntimeItem> =
        RuntimeEventSchema.visibleTranscriptItems(items)

    fun openRequestsFromRuntimeItems(
        items: List<PersistedRuntimeItem>,
        threadId: String,
        nowEpochMs: Long = System.currentTimeMillis(),
    ): List<OpenRuntimeRequest> =
        RuntimeEventSchema.openRequestsFromRuntimeItems(items, threadId, nowEpochMs)


    const val PENDING_REQUEST_ITEM_TYPE = RuntimeEventSchema.PENDING_REQUEST_ITEM_TYPE

    // MARK: - Item + domain reduction

    private val stateRank = mapOf(
        "started" to 0,
        "updated" to 1,
        "completed" to 2,
    )

    fun apply(events: List<RuntimeEvent>, items: MutableList<PersistedRuntimeItem>) {
        for (event in events) {
            apply(event, items)
        }
    }

    fun apply(event: RuntimeEvent, items: MutableList<PersistedRuntimeItem>) {
        when (event.type) {
            "item.started" -> {
                val itemId = event.itemId ?: return
                if (items.any { it.id == itemId }) return
                items.add(
                    PersistedRuntimeItem(
                        id = itemId,
                        type = event.itemType ?: "unknown",
                        state = "started",
                        payload = event.payload,
                        streams = emptyMap(),
                        parentItemId = event.parentItemId,
                    ),
                )
            }

            "item.updated" -> {
                // Requires payloadSpecified (enforced in parse); never fabricate missing item.
                val itemId = event.itemId ?: return
                val index = items.indexOfFirst { it.id == itemId }
                if (index < 0) return
                val item = items[index]
                // TS always mergePayload(prev, event.payload). When key present
                // (including null), apply; when absent this event would not parse.
                val nextPayload = if (event.payloadSpecified) {
                    mergePayload(item.payload, event.payload.takeUnless { it is JsonNull })
                } else {
                    // Should not reach here if parse enforced presence.
                    item.payload
                }
                items[index] = item.copy(
                    state = monotonicState(item.state, "updated"),
                    payload = nextPayload,
                )
            }

            "item.completed" -> {
                val itemId = event.itemId ?: return
                val index = items.indexOfFirst { it.id == itemId }
                if (index < 0) return
                val item = items[index]
                val nextPayload = if (event.payloadSpecified) {
                    mergePayload(item.payload, event.payload.takeUnless { it is JsonNull })
                } else {
                    item.payload
                }
                val next = item.copy(
                    state = "completed",
                    payload = nextPayload,
                )
                if (next.type == "reasoning" &&
                    next.streams["reasoning_text"].orEmpty().trim().isEmpty()
                ) {
                    items.removeAt(index)
                    return
                }
                items[index] = next
            }

            "content.delta" -> {
                val itemId = event.itemId ?: return
                val stream = event.stream ?: return
                val delta = event.delta ?: return
                val index = items.indexOfFirst { it.id == itemId }
                if (index < 0) return
                val item = items[index]
                val streams = item.streams.toMutableMap()
                streams[stream] = (streams[stream].orEmpty()) + delta
                items[index] = item.copy(
                    streams = streams,
                    state = monotonicState(item.state, "updated"),
                )
            }

            "error" -> {
                val id = "err-${UUID.randomUUID()}"
                val message = event.message
                    ?: event.raw.string("message")
                    ?: "Runtime error"
                items.add(
                    PersistedRuntimeItem(
                        id = id,
                        type = "error",
                        state = "completed",
                        payload = buildJsonObject { put("message", message) },
                        streams = emptyMap(),
                        parentItemId = null,
                    ),
                )
            }

            "turn.completed" -> {
                val state = event.state ?: event.raw.string("state")
                if (state == "interrupted" || state == "cancelled") {
                    pruneTrailingInterruptedReasoning(items)
                }
            }

            else -> {
                // Request/turn/context events are handled in the domain reducer;
                // warning and usage.spent are intentional renderer no-ops.
            }
        }
    }

    /**
     * Full apply: items + domain state for one batch.
     * [uuidClock] injects deterministic ids/times in tests.
     */
    fun applyBatch(
        events: List<RuntimeEvent>,
        threadId: String,
        items: MutableList<PersistedRuntimeItem>,
        domain: ThreadRuntimeDomainState,
        nowEpochMs: Long = System.currentTimeMillis(),
    ): ThreadRuntimeDomainState {
        var nextDomain = domain
        for (event in events) {
            apply(event, items)
            nextDomain = RuntimeDomainReducer.apply(event, threadId, nextDomain, nowEpochMs)
        }
        return nextDomain
    }

    /**
     * Trailing reasoning after interrupted/cancelled turn is dropped.
     * Plan / error / parented (child) items are transparent — skipped, not dropped.
     */
    fun pruneTrailingInterruptedReasoning(items: MutableList<PersistedRuntimeItem>) {
        val dropIds = mutableSetOf<String>()
        var idx = items.lastIndex
        while (idx >= 0) {
            val item = items[idx]
            if (item.type == "plan" || item.type == "error" || item.parentItemId != null) {
                idx -= 1
                continue
            }
            if (item.type != "reasoning") break
            dropIds.add(item.id)
            idx -= 1
        }
        if (dropIds.isNotEmpty()) {
            items.removeAll { it.id in dropIds }
        }
    }

    fun shouldRefreshShell(from: JsonElement): Boolean {
        val type = from.asObjectOrNull()?.string("type") ?: return false
        if (isShellLifecycleType(type)) return true
        // Probe nested event type strings without requiring full schema validity
        // so incomplete envelopes still schedule shell refresh.
        for (nestedType in nestedEventTypes(from)) {
            if (isShellLifecycleType(nestedType)) return true
        }
        return false
    }

    fun shouldRefreshOpenThreadMetadata(from: JsonElement): Boolean {
        val objectMap = from.asObjectOrNull() ?: return false
        val type = objectMap.string("type") ?: return false
        if (isOpenThreadMetadataType(type)) return true
        for (nestedType in nestedEventTypes(from)) {
            if (isOpenThreadMetadataType(nestedType)) return true
        }
        return false
    }

    /** Collect nested runtime event type discriminators (best-effort, non-strict). */
    private fun nestedEventTypes(from: JsonElement): List<String> {
        val objectMap = from.asObjectOrNull() ?: return emptyList()
        return when (objectMap.string("type")) {
            "thread-runtime-event" -> {
                listOfNotNull(objectMap.obj("event")?.string("type"))
            }
            "thread-runtime-events" -> {
                objectMap.array("events")?.mapNotNull { it.asObjectOrNull()?.string("type") }
                    .orEmpty()
            }
            "thread-runtime-events-multi" -> {
                objectMap.array("batches")?.flatMap { batch ->
                    batch.asObjectOrNull()?.array("events")?.mapNotNull {
                        it.asObjectOrNull()?.string("type")
                    }.orEmpty()
                }.orEmpty()
            }
            else -> emptyList()
        }
    }

    private fun isShellLifecycleType(type: String): Boolean {
        if (type == "remote-projects-changed" ||
            type == "remote-threads-changed" ||
            type == "thread-state"
        ) {
            return true
        }
        // content.delta alone is not a shell lifecycle signal (matches prior tests).
        if (type == "content.delta") return false
        return type.startsWith("turn.") ||
            type.startsWith("session.") ||
            type.startsWith("item.") ||
            type.startsWith("request.") ||
            type == "error" ||
            type == "warning"
    }

    private fun isOpenThreadMetadataType(type: String): Boolean {
        if (type == "thread-state" ||
            type.startsWith("turn.") ||
            type.startsWith("session.") ||
            type == "error" ||
            type == "warning" ||
            type.startsWith("request.")
        ) {
            return true
        }
        return false
    }

    private fun monotonicState(current: String, incoming: String): String {
        if (current == "completed") return "completed"
        val currentRank = stateRank[current] ?: 0
        val incomingRank = stateRank[incoming] ?: 0
        return if (incomingRank >= currentRank) incoming else current
    }

    /**
     * Shallow-merge object payloads (TS `mergePayload`):
     * - [incoming] null → null (clear) when caller intends replace
     * - non-object either side → return [incoming] wholesale
     * - both objects → top-level keys from [incoming] replace; nested not recursive
     *
     * Note: for `item.updated` with payloadSpecified and JsonNull, caller passes null
     * after takeUnless so payload clears (TS mergePayload returns next when non-object).
     */
    fun mergePayload(existing: JsonElement?, incoming: JsonElement?): JsonElement? {
        if (incoming == null) return null
        if (incoming is JsonNull) return null
        if (existing == null) return incoming
        val left = existing.asObjectOrNull()
        val right = incoming.asObjectOrNull()
        if (left == null || right == null) return incoming
        val merged = left.toMutableMap()
        for ((key, value) in right) {
            merged[key] = value
        }
        return JsonObject(merged)
    }

    val CANONICAL_TRANSCRIPT_STREAMS: List<String> = TranscriptText.preferredStreamKeys

    fun extractTranscriptText(item: PersistedRuntimeItem): String = item.displayText

    fun contentDeltaEvent(
        threadId: String?,
        itemId: String,
        stream: String,
        delta: String,
        parentItemId: String? = null,
        raw: JsonObject = JsonObject(emptyMap()),
    ): RuntimeEvent = RuntimeEvent(
        type = "content.delta",
        threadId = threadId,
        itemId = itemId,
        stream = stream,
        delta = delta,
        parentItemId = parentItemId,
        raw = raw,
    )
}

File: ios/App/App/Protocol/RuntimeEventReducer.swift
import Foundation

/// Unwraps remote supervisor envelopes and reduces runtime item events for a thread transcript.
/// Aligns with `src/renderer/state/slices/runtimeEventReducer.ts` item/delta semantics:
/// - `item.started` creates; `item.updated` / `item.completed` / `content.delta` never fabricate
/// - payload merge is **shallow** top-level field replace (not recursive nested retention)
/// - `item.updated` **requires** payload key (strict decoder); explicit null clears
/// - `item.completed` with empty reasoning_text deletes the item
/// - `error` synthesizes a completed error item
/// - `turn.completed` interrupted/cancelled prunes trailing reasoning
/// - `request.opened` / `request.resolved` tracked in domain request state
/// - `usage.spent` is a no-op (main-process ledger only; no native ledger here)
enum RuntimeEventReducer {
    struct RuntimeEvent: Sendable, Equatable {
        var type: String
        var threadId: String?
        var itemId: String?
        var itemType: String?
        var state: String?
        var stream: String?
        var delta: String?
        var payload: JSONValue?
        /// True when the wire object contained a `payload` key (even if null).
        var payloadSpecified: Bool
        var parentItemId: String?
        var requestId: String?
        var requestType: String?
        var message: String?
        var turnId: String?
        var contextUsage: ThreadContextUsage?
        var raw: [String: JSONValue]

        /// Test/helper convenience: `payloadSpecified` defaults to `payload != nil`.
        init(
            type: String,
            threadId: String? = nil,
            itemId: String? = nil,
            itemType: String? = nil,
            state: String? = nil,
            stream: String? = nil,
            delta: String? = nil,
            payload: JSONValue? = nil,
            payloadSpecified: Bool? = nil,
            parentItemId: String? = nil,
            requestId: String? = nil,
            requestType: String? = nil,
            message: String? = nil,
            turnId: String? = nil,
            contextUsage: ThreadContextUsage? = nil,
            raw: [String: JSONValue] = [:]
        ) {
            self.type = type
            self.threadId = threadId
            self.itemId = itemId
            self.itemType = itemType
            self.state = state
            self.stream = stream
            self.delta = delta
            self.payload = payload
            self.payloadSpecified = payloadSpecified ?? (payload != nil)
            self.parentItemId = parentItemId
            self.requestId = requestId
            self.requestType = requestType
            self.message = message
            self.turnId = turnId
            self.contextUsage = contextUsage
            self.raw = raw
        }
    }

    struct Batch: Sendable, Equatable {
        var threadId: String
        var events: [RuntimeEvent]
    }

    /// Open request tracked from request.opened until request.resolved.
    struct OpenRuntimeRequest: Sendable, Equatable, Identifiable {
        var id: String { requestId }
        var requestId: String
        var threadId: String
        var requestType: String?
        var payload: JSONValue?
        var receivedAt: Date
    }

    // MARK: - Envelope unwrapping

    /// Collects runtime event batches from `thread-runtime-event`,
    /// `thread-runtime-events`, and `thread-runtime-events-multi` envelopes.
    static func collectRuntimeEvents(from supervisory: JSONValue) -> [Batch] {
        guard case .object(let object) = supervisory,
              let type = object["type"]?.stringValue
        else {
            return []
        }

        switch type {
        case "thread-runtime-event":
            guard let threadId = object["threadId"]?.stringValue,
                  let eventValue = object["event"],
                  case .object(let eventObject) = eventValue,
                  let event = RuntimeEventDecoder.decode(eventObject)
            else {
                return []
            }
            return [Batch(threadId: threadId, events: [event])]

        case "thread-runtime-events":
            guard let threadId = object["threadId"]?.stringValue,
                  case .array(let rawEvents) = object["events"]
            else {
                return []
            }
            let events = rawEvents.compactMap { value -> RuntimeEvent? in
                guard case .object(let obj) = value else { return nil }
                return RuntimeEventDecoder.decode(obj)
            }
            return events.isEmpty ? [] : [Batch(threadId: threadId, events: events)]

        case "thread-runtime-events-multi":
            guard case .array(let batches) = object["batches"] else { return [] }
            var results: [Batch] = []
            for batchValue in batches {
                guard case .object(let batchObject) = batchValue,
                      let threadId = batchObject["threadId"]?.stringValue,
                      case .array(let rawEvents) = batchObject["events"]
                else {
                    continue
                }
                let events = rawEvents.compactMap { value -> RuntimeEvent? in
                    guard case .object(let obj) = value else { return nil }
                    return RuntimeEventDecoder.decode(obj)
                }
                if !events.isEmpty {
                    results.append(Batch(threadId: threadId, events: events))
                }
            }
            return results

        default:
            return []
        }
    }

    // MARK: - Item reduction

    private static let stateRank: [String: Int] = [
        "started": 0,
        "updated": 1,
        "completed": 2,
    ]

    /// Applies a batch of runtime events to an item list (open-thread only).
    static func apply(
        events: [RuntimeEvent],
        to items: inout [PersistedRuntimeItem]
    ) {
        for event in events {
            apply(event: event, to: &items)
        }
    }

    /// Apply item-mutating events. Domain fields (open-turn/requests/context) via `applyDomain`.
    static func apply(event: RuntimeEvent, to items: inout [PersistedRuntimeItem]) {
        switch event.type {
        case "item.started":
            guard let itemId = event.itemId else { return }
            if items.contains(where: { $0.id == itemId }) { return }
            items.append(
                PersistedRuntimeItem(
                    id: itemId,
                    type: event.itemType ?? "unknown",
                    state: "started",
                    payload: event.payload,
                    streams: [:],
                    parentItemId: event.parentItemId
                )
            )

        case "item.updated":
            // Strict: payload key required (decoder rejects absent). Direct apply also no-ops.
            guard event.payloadSpecified,
                  let itemId = event.itemId,
                  let index = items.firstIndex(where: { $0.id == itemId })
            else { return }
            var item = items[index]
            item.state = monotonicState(current: item.state, incoming: "updated")
            // Explicit null clears; object shallow-merges.
            item.payload = mergePayload(item.payload, event.payload)
            items[index] = item

        case "item.completed":
            guard let itemId = event.itemId,
                  let index = items.firstIndex(where: { $0.id == itemId })
            else { return }
            var item = items[index]
            item.state = "completed"
            // Absent payload retains; explicit null clears; object merges.
            if event.payloadSpecified {
                item.payload = mergePayload(item.payload, event.payload)
            }
            if item.type == "reasoning" {
                let text = (item.streams["reasoning_text"] ?? "").trimmingCharacters(
                    in: .whitespacesAndNewlines
                )
                if text.isEmpty {
                    items.remove(at: index)
                    return
                }
            }
            items[index] = item

        case "content.delta":
            guard let itemId = event.itemId,
                  let stream = event.stream,
                  let delta = event.delta,
                  let index = items.firstIndex(where: { $0.id == itemId })
            else {
                return
            }
            var item = items[index]
            var streams = item.streams
            streams[stream, default: ""] += delta
            item.streams = streams
            item.state = monotonicState(current: item.state, incoming: "updated")
            items[index] = item

        case "error":
            let id = "err-\(UUID().uuidString)"
            let message = event.message
                ?? event.raw["message"]?.stringValue
                ?? "Runtime error"
            items.append(
                PersistedRuntimeItem(
                    id: id,
                    type: "error",
                    state: "completed",
                    payload: .object(["message": .string(message)]),
                    streams: [:],
                    parentItemId: nil
                )
            )

        case "turn.completed":
            let state = event.state ?? event.raw["state"]?.stringValue
            if state == "interrupted" || state == "cancelled" {
                pruneTrailingInterruptedReasoning(from: &items)
            }

        default:
            // session/warning/usage/turn.started/request.*/context.updated do not mutate items.
            break
        }
    }

    /// Apply domain mutations (open-turn, requests, context usage, completed turns).
    /// `usage.spent` is intentionally a no-op (no native usage ledger).
    static func applyDomain(
        event: RuntimeEvent,
        threadId: String,
        domain: inout RuntimeThreadDomainState
    ) {
        switch event.type {
        case "turn.started":
            if domain.openTurn != true {
                domain.openTurn = true
            }

        case "turn.completed":
            domain.openTurn = false
            if let turnId = event.turnId, let state = event.state {
                domain.completedTurns.append(
                    CompletedTurnRecord(turnId: turnId, state: state)
                )
            }
            if eventAffectsStructuralVersion(event) {
                domain.structuralVersion += 1
            }

        case "context.updated":
            if let usage = event.contextUsage {
                domain.contextUsage = usage
            }

        case "usage.spent":
            // Main-process ledger only — no conflicting native consumer.
            break

        case "request.opened":
            guard let requestId = event.requestId else { return }
            domain.openRequests.removeAll { $0.requestId == requestId }
            domain.openRequests.append(
                OpenRuntimeRequest(
                    requestId: requestId,
                    threadId: threadId,
                    requestType: event.requestType,
                    payload: event.payload,
                    receivedAt: Date()
                )
            )

        case "request.resolved":
            guard let requestId = event.requestId else { return }
            domain.openRequests.removeAll { $0.requestId == requestId }

        case "item.started", "item.updated", "item.completed", "error":
            if eventAffectsStructuralVersion(event) {
                domain.structuralVersion += 1
            }

        default:
            break
        }
    }

    /// Track request.opened / request.resolved (legacy entry used by session request list).
    static func applyRequestEvent(
        event: RuntimeEvent,
        threadId: String,
        to requests: inout [OpenRuntimeRequest]
    ) {
        var domain = RuntimeThreadDomainState(openRequests: requests)
        applyDomain(event: event, threadId: threadId, domain: &domain)
        requests = domain.openRequests
    }

    /// Hydrate domain from a history snapshot (contextUsage, completedTurns, pending requests).
    static func hydrateDomain(
        from history: RemoteThreadSnapshot,
        into domain: inout RuntimeThreadDomainState
    ) {
        // Install snapshot.contextUsage when valid.
        if let raw = history.contextUsage, case .object(let obj) = raw,
           let usage = RuntimeEventDecoder.decodeContextUsage(obj) {
            domain.contextUsage = usage
        }
        // completedTurns from snapshot when present as objects with turnId/state.
        var turns: [CompletedTurnRecord] = []
        for value in history.completedTurns {
            guard case .object(let obj) = value,
                  let turnId = obj["turnId"]?.stringValue ?? obj["id"]?.stringValue,
                  let state = obj["state"]?.stringValue
            else { continue }
            turns.append(CompletedTurnRecord(turnId: turnId, state: state))
        }
        if !turns.isEmpty {
            domain.completedTurns = turns
        }

        // Derive still-open requests from started/non-completed `pending_request` items
        // only when thread status requires approval/reply. Outer payload carries requestId;
        // use that (not item id), validate canonical inner payload, dedupe by requestId
        // preserving last write order (FIFO map overwrite).
        let status = history.thread.status
        if status == "needs_approval" || status == "needs_reply" {
            var ordered: [String: OpenRuntimeRequest] = [:]
            var order: [String] = []
            for item in history.runtimeItems {
                guard item.type == "pending_request", item.state != "completed" else { continue }
                guard let decoded = RuntimeEventDecoder.decodePendingRequestOuter(item.payload)
                else { continue }
                let request = OpenRuntimeRequest(
                    requestId: decoded.requestId,
                    threadId: history.thread.id,
                    requestType: decoded.requestType,
                    payload: decoded.payload,
                    receivedAt: Date()
                )
                if ordered[decoded.requestId] == nil {
                    order.append(decoded.requestId)
                }
                ordered[decoded.requestId] = request
            }
            domain.openRequests = order.compactMap { ordered[$0] }
        } else {
            domain.openRequests = []
        }
    }

    static func pruneTrailingInterruptedReasoning(from items: inout [PersistedRuntimeItem]) {
        var dropIds = Set<String>()
        var idx = items.count - 1
        while idx >= 0 {
            let item = items[idx]
            if item.type == "plan" || item.type == "error" || item.parentItemId != nil {
                idx -= 1
                continue
            }
            if item.type != "reasoning" { break }
            dropIds.insert(item.id)
            idx -= 1
        }
        if !dropIds.isEmpty {
            items.removeAll { dropIds.contains($0.id) }
        }
    }

    static func shouldRefreshShell(from supervisory: JSONValue) -> Bool {
        guard case .object(let object) = supervisory,
              let type = object["type"]?.stringValue
        else {
            return false
        }
        if isShellLifecycleType(type) {
            return true
        }
        let batches = collectRuntimeEvents(from: supervisory)
        for batch in batches {
            for event in batch.events {
                if isShellLifecycleType(event.type) {
                    return true
                }
            }
        }
        return false
    }

    static func shouldRefreshOpenThreadMetadata(from supervisory: JSONValue) -> Bool {
        guard case .object(let object) = supervisory,
              let type = object["type"]?.stringValue
        else {
            return false
        }
        if isOpenThreadMetadataType(type) {
            return true
        }
        let batches = collectRuntimeEvents(from: supervisory)
        for batch in batches {
            for event in batch.events {
                if isOpenThreadMetadataType(event.type) {
                    return true
                }
            }
        }
        return false
    }

    private static func isShellLifecycleType(_ type: String) -> Bool {
        if type == "remote-projects-changed"
            || type == "remote-threads-changed"
            || type == "thread-state" {
            return true
        }
        return type.hasPrefix("turn.") || type.hasPrefix("session.")
    }

    private static func isOpenThreadMetadataType(_ type: String) -> Bool {
        if type == "thread-state"
            || type.hasPrefix("turn.")
            || type.hasPrefix("session.") {
            return true
        }
        if type == "error" || type == "warning" || type.hasPrefix("request.") {
            return true
        }
        return false
    }

    private static func eventAffectsStructuralVersion(_ event: RuntimeEvent) -> Bool {
        switch event.type {
        case "item.started", "item.updated", "item.completed", "turn.completed", "error":
            return true
        default:
            return false
        }
    }

    // MARK: - Helpers

    private static func monotonicState(current: String, incoming: String) -> String {
        let currentRank = stateRank[current] ?? 0
        let incomingRank = stateRank[incoming] ?? 0
        if current == "completed" { return "completed" }
        return incomingRank >= currentRank ? incoming : current
    }

    /// Shallow-merge object payloads (TS `mergePayload`):
    /// - nil / JSON null / non-object `incoming` replaces entirely (including clear)
    /// - object+object: top-level keys from incoming replace; nested not recursive
    static func mergePayload(_ existing: JSONValue?, _ incoming: JSONValue?) -> JSONValue? {
        guard let incoming else { return nil }
        if case .null = incoming { return nil }
        guard case .object = incoming else { return incoming }
        guard let existing, case .object(let left) = existing,
              case .object(let right) = incoming
        else {
            return incoming
        }
        var merged = left
        for (key, value) in right {
            merged[key] = value
        }
        return .object(merged)
    }
}
